### ISSUE TITLE
re \w does not match some valid Unicode characters

### ISSUE THREAD

**Issue by e98a6e43-25c8-4e1f-8eba-4b9486e14aef** [python/cpython#75021]
BPO | [30838](https://bugs.python.org/issue30838)
--- | :---
Nosy | @vstinner, @ezio-melotti, @serhiy-storchaka, @ThiefMaster, @davidism

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-07-05.15:19:42.232>
created_at = <Date 2017-07-03.15:39:29.866>
labels = ['expert-regex', 'invalid', 'type-bug', '3.7', 'expert-unicode']
title = 're \\w does not match some valid Unicode characters'
updated_at = <Date 2017-07-05.17:22:30.253>
user = 'https://github.com/davidism'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-05.17:22:30.253>
actor = 'mrabarnett'
assignee = 'none'
closed = True
closed_date = <Date 2017-07-05.15:19:42.232>
closer = 'davidism'
components = ['Regular Expressions', 'Unicode']
creation = <Date 2017-07-03.15:39:29.866>
creator = 'davidism'
dependencies = []
files = []
hgrepos = []
issue_num = 30838
keywords = []
message_count = 5.0
messages = ['297603', '297604', '297613', '297766', '297773']
nosy_count = 6.0
nosy_names = ['vstinner', 'ezio.melotti', 'mrabarnett', 'serhiy.storchaka', 'ThiefMaster', 'davidism']
pr_nums = []
priority = 'normal'
resolution = 'not a bug'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue30838'
versions = ['Python 3.3', 'Python 3.4', 'Python 3.5', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by e98a6e43-25c8-4e1f-8eba-4b9486e14aef** [comment]
This came up while writing a regex to match characters that are valid in Python identifiers for Jinja. https://github.com/pallets/jinja/pull/731 `\w` matches all valid identifier characters except for 4 special cases:

```py
import unicodedata
import re
import sys

cre = re.compile(r'\w')

for cp in range(sys.maxunicode + 1):
    s = chr(cp)

    if s.isidentifier() and not cre.match(s):
        print(hex(cp), unicodedata.name(s))
```


0x1885 MONGOLIAN LETTER ALI GALI BALUDA
0x1886 MONGOLIAN LETTER ALI GALI THREE BALUDA
0x2118 SCRIPT CAPITAL P
0x212e ESTIMATED SYMBOL

Python \< 3.6 matches the two Mongolian characters, not sure why 3.6 stopped matching them.

For our case, we just added them to a character set, `[\w\u1885\u1886\u2118\u212e]`.

It can cause unexpected behavior when using `\b`, since that's defined as the transition from `\w` to `\W` and those 4 characters aren't in `\w`. `re.match(r'\b[\w\u212e', '℮')` fails to match.

**Comment by e98a6e43-25c8-4e1f-8eba-4b9486e14aef** [comment]
Adding `or ('a' + s).isidentifer()`, to catch valid id_continue characters, to the test in the previous script reveals many more characters that seem like valid word characters but aren't matched by `\w`.

**Comment by 39d85a87-36ea-41b2-b2bb-2be43abb500e** [comment]
In Unicode 9.0.0, U+1885 and U+1886 changed from being General_Category=Other_Letter (Lo) to General_Category=Nonspacing_Mark (Mn).

U+2118 is General_Category=Math_Symbol (Sm) and U+212E is General_Category=Other_Symbol (So).

\\w doesn't include Mn, Sm or So.

The .identifier method uses the Unicode properties XID_Start and XID_Continue, which include these codepoints.

**Comment by e98a6e43-25c8-4e1f-8eba-4b9486e14aef** [comment]
After thinking about it more, I guess I misunderstood what \\w was doing compared to isidentifier. Since Python just relies on the Unicode database, there's not much to be done anyway. Closing this.

For anyone interested, we ended up with a hybrid approach for lexing identifiers: build a regex group that includes all valid ranges not matched by \\w, then validate with isidentifier later. https://github.com/pallets/jinja/pull/731/files
